### ISSUE TITLE
connectionLost callback not triggered for async client

### ISSUE THREAD

**Issue by SantinoKeupp** [eclipse-paho/paho.mqtt.c#974]
Hello,

I am originally coming from the paho mqtt cpp lib and tracked my problem down here. It looks like the `MQTTAsync_connectionLost` callback will not be triggered if the broker (in my case mosquitto) goes down.

I can reproduce this with a slightly modified MQTTAsync_subscribe sample program. I just changed the defined `ADDRESS` to "tcp://localhost:1883". When killing the broker (mosquitto), the ~~`MQTTAsync_subscribe`~~ `MQTTAsync_connectionLost` callback will not be called (Instead  ~~`onConnectFailure`~~ `MQTTAsync_onFailure` will be called, if set).
If I do the same with the MQTTClient_subscribe sample program, everything works as expected.

Am I missing something or is this a bug?

Thanks,
Santino

**Comment by SantinoKeupp** [comment]
OS: Linux (gentoo)
version: 1.3.5(originally noticed) / master - e4c3757 (reproduced)

**Comment by icraggs** [comment]
I'm a bit confused.  What callback isn't called - connectionLost or the MQTTAsync_subscribe one?  You're not going to get the connectionLost callback unless the connect has succeeded, which if onConnectFailure has been called, it hasn't.

**Comment by SantinoKeupp** [comment]
Thank you for your quick response and sorry for the confusing copy and past error. I edited my original post. I am not getting the connectionLost callback, but the onConnectFailure even after the connection has been established and i successfully received messages. But as I said, only on the asyn client:

``` bash
$ ./MQTTAsync_subscribe
Successful connection
Subscribing to topic MQTT Examples
for client ExampleClientSub using QoS1

Press Q<Enter> to quit

Subscribe succeeded
Message arrived
     topic: MQTT Examples
   message: hello world
Connect failed, rc -11
``` 
With the synchronous client it looks like this:
``` bash
$ ./MQTTClient_subscribe                          
Subscribing to topic MQTT Examples
for client ExampleClientSub using QoS1

Press Q<Enter> to quit

Message arrived
     topic: MQTT Examples
   message: hello world

Connection lost
     cause: (null)
 ```

**Comment by icraggs** [comment]
Introduced while fixing #916.  I tried a fix in v1.3.6 but that apparently wasn't correct.  The fix in the develop branch works for me.

**Comment by SantinoKeupp** [comment]
Works for me as well, but without further testing. Thank you!
